File: common/src/camera.rs
use crate::ray::Ray;
use crate::rtweekend::{degrees_to_radians, random_in_unit_disk};
use crate::vec3::Vec3;
use crate::Point3;

pub struct Camera {
    pub origin: Point3,
    pub lower_left_corner: Point3,
    pub horizontal: Vec3,
    pub vertical: Vec3,
    pub u: Vec3, // horizontal
    pub v: Vec3, // vertical
    pub w: Vec3, // depth
    pub lens_radius: f64,
}

impl Camera {
    /// lookfrom: the position where we place the camera
    /// lookat: the point we look at
    /// vup: view up vector
    /// vfov: vertical field-of-view in degrees
    /// aspect_ration: width / height
    pub fn new(
        lookfrom: Point3,
        lookat: Point3,
        vup: Vec3,
        vfov: f64,
        aspect_ratio: f64,
        aperture: f64,
        focus_dist: f64,
    ) -> Self {
        let theta = degrees_to_radians(vfov);
        let h = (theta / 2.0).tan();
        let viewport_height = 2.0 * h;
        let viewport_width = aspect_ratio * viewport_height;

        let w = (lookfrom - lookat).unit_vector();
        let u = vup.cross(&w).unit_vector();
        let v = w.cross(&u);

        let origin = lookfrom;
        let horizontal = focus_dist * viewport_width * u;
        let vertical = focus_dist * viewport_height * v;
        let lower_left_corner = origin - horizontal / 2.0 - vertical / 2.0 - focus_dist * w;

        let lens_radius = aperture / 2.0;
        Camera {
            origin: origin.clone(),
            lower_left_corner,
            horizontal,
            vertical,
            u,
            v,
            w,
            lens_radius,
        }
    }

    pub fn get_ray(&self, s: f64, t: f64) -> Ray {
        let rd = self.lens_radius * random_in_unit_disk();
        let offset = self.u * rd.x + self.v * rd.y;
        Ray {
            origin: self.origin + offset,
            direction: self.lower_left_corner + s * self.horizontal + t * self.vertical
                - self.origin
                - offset,
        }
    }
}

File: tj/camera.py
import math

import taichi as ti
from taichi.math import *

from ray import Ray
from rtweekend import random_in_unit_disk


@ti.data_oriented
class Camera:
    origin: vec3
    horizontal: vec3
    vertical: vec3
    lower_left_corner: vec3
    len_radius: float
    w: vec3
    u: vec3
    v: vec3

    def __init__(self,
                 lookfrom: vec3,
                 lookat: vec3,
                 vup: vec3,
                 vfov: float,
                 aspect_ratio: float,
                 aperture: float,
                 focus_dist: float):
        theta = vfov * math.pi / 180
        h = ti.tan(theta / 2)
        viewport_height = 2.0 * h
        viewport_width = aspect_ratio * viewport_height

        self.w = normalize(lookfrom - lookat)
        self.u = normalize(cross(vup, self.w))
        self.v = cross(self.w, self.u)

        self.origin = lookfrom
        self.horizontal = focus_dist * viewport_width * self.u
        self.vertical = focus_dist * viewport_height * self.v
        self.lower_left_corner = vec3(
            self.origin - self.horizontal / 2 - self.vertical / 2 - focus_dist * self.w)
        self.len_radius = aperture / 2

    @ti.func
    def get_ray(self, u: float, v: float) -> Ray:
        rd = self.len_radius * random_in_unit_disk()
        offset = rd[0] * self.u + rd[1] * self.v
        return Ray(self.origin + offset,
                   self.lower_left_corner + u * self.horizontal + v * self.vertical - self.origin - offset)
